### ISSUE TITLE
Problem with future forecasting results (Unreasonable forecasting when assigning 2070 variable)

### ISSUE THREAD

**Issue by joshikp01** [biomodhub/biomod2#467]
Hello !!
I am trying to use biomod2 to see the range changes in current and in 2070 for a tree species but the results are showing very unreasonable such that no suitable range will be left by 2070. I need assistance on this matter. Please note that my variables have the same extent and resolutions and when I build the model with future projected data from the start the model runs fine and there is very good amount of suitable range.
![problembiomod2](https://github.com/biomodhub/biomod2/assets/167947828/c38a1e54-3707-4ec7-9234-262908cad209)

(Note: My problem is very similar to issue #426 but i could not find the solution).

My code is given below: (I have also attached my files (environmental variables and presence data below)

# Load required packages
library("raster")
library("ggplot2")
library("gridExtra")
library("rasterVis")
library("sf")
library("usdm")
library("biomod2")
library("dismo")
packageVersion("biomod2") #it is 4.2.5

# List of file paths (makes the file in same extent)
file_paths <- c(
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/dem.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/slope.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/aspect.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio3.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio4.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio12.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio14.asc",
  "F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio15.asc"
)

# Load rasters
rasters <- lapply(file_paths, raster)

# Resample rasters to match bio1
rasters <- lapply(rasters, function(r) {
  resample(r, bio3)
})

# Overwrite existing raster files
for (i in seq_along(rasters)) {
  writeRaster(rasters[[i]], filename = file_paths[i], overwrite = TRUE)
}

# Check for warnings
warnings()
#variables for current projection:

DEM<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/dem.asc")
slope<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/slope.asc")
aspect<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/aspect.asc")
bio3<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio3.asc")
bio4<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio4.asc")
bio12<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio12.asc")
bio14<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio14.asc")
bio15<- raster ("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/present/asc/bio15.asc")

#rasterstack of layers with below 5 VIF and 0.7 Corr
myExpl1<- raster::stack (DEM, slope, aspect, bio3, bio4, bio12, bio14, bio15)
summary(myExpl1)

thinned_sal<- read.csv("F:/Khagendra (Ensemble analysis)/R/inventory data/sal.csv")

#Species data 
myRespName <- 'Sal'
myRespName

# My Resp Data 
myResp <- as.numeric(thinned_sal [,myRespName])
myResp

head(myResp)
# The XY coordinates of species data 
myRespXY <- thinned_sal [,c("Longitude", "Latitude")] 
n=round(nrow(thinned_sal)*0.8) #(counts the number of rows of colum to fewed it into the downsampled RF)

myRespXY
?BIOMOD_FormatingData
# Formating biomod data
myBiomodData <- BIOMOD_FormatingData(resp.var = myResp, 
                                     expl.var = myExpl1, 
                                     resp.xy = myRespXY, 
                                     resp.name = myRespName, 
                                     PA.nb.rep = 3, 
                                     PA.nb.absences = 10000,
                                     PA.strategy = 'random') 
?BIOMOD_FormatingData
myBiomodData
#Define RF options
myRFoptions <- list(ntree=500, sampsize =c("0"=n, "1"= n), replace=TRUE, nodesize=5)
user.RF <- rep( list(myRFoptions), (ncol(myBiomodData@PA.table) + 1 ))
names(user.RF) <- c( paste0("_", names(myBiomodData@PA.table), "_allRun"), "_allData_allRun")

#Define user values
user.val <- list(RF.binary.randomForest.randomForest = user.RF)

all.models<- c("GAM", "GBM", "GLM", "MAXENT", "RF")  

#Set up modeling options
myBiomodOption <- bm_ModelingOptions(
  data.type = "binary",
  models = all.models,
  strategy = 'user.defined',
  user.val= user.val,
  bm.format = myBiomodData,
  calib.lines = NULL
)

myBiomodModelOut <- BIOMOD_Modeling(
  bm.format = myBiomodData,
  modeling.id = paste(myRespName, "Sal", sep = "_"),
  models = all.models,
  OPT.strategy= 'user.defined',
  OPT.user = myBiomodOption,  #!!
  CV.strategy = "random", 
  CV.nb.rep = 3,
  CV.perc = 0.8, 
  metric.eval = c("ROC", "TSS"), 
  var.import = 0, 
  scale.models = FALSE, 
  nb.cpu = 8, 
  do.progress = TRUE, 
  seed.val = 42,
)

myCalibLines <- get_calib_lines(myBiomodModelOut)
plot(myBiomodData, calib.lines = myCalibLines)

# Get all models evaluation
myBiomodModelEval <- get_evaluations(myBiomodModelOut)
myBiomodModelEval
export<- as.data.frame(myBiomodModelEval)

write.csv(export, "F:/Khagendra (Ensemble analysis)/fig/ensembelresults.csv")

# Plot model evaluation scores

plot<- bm_PlotEvalMean(myBiomodModelOut, metric.eval = c("ROC", "TSS"), dataset="validation", group.by = "algo", do.plot = TRUE) #if you wanna assign to plot

plot

# Select the best performing model 
myModel.resp <- get_built_models(myBiomodModelOut, algo = c('RF', 'GBM'))
?BIOMOD_LoadModels
# Plot the response curve for selected model 

?bm_PlotEvalBoxplot
?bm_PlotResponseCurves
myRespPlot2D <- bm_PlotResponseCurves(
  models = myModel.resp,
  Data = get_formal_data(myBiomodModelOut, "expl.var"),
  show.variables = get_formal_data(myBiomodModelOut, "expl.var.names"),
  do.bivariate = FALSE,
  fixed.var.metric = "mean",
  group.by="algo",
  col = c("blue", "red"),
  data_species = get_formal_data(myBiomodModelOut, "resp.var"),
  bm.out = myBiomodModelOut
)
?BIOMOD_EnsembleModeling
myBiomodEM <- BIOMOD_EnsembleModeling(
  bm.mod =myBiomodModelOut,
  models.chosen = 'all',
  em.by='all',
  em.algo = c( 'EMwmean'),
  metric.select = c('ROC'),
  metric.select.thresh= c(0.8),
  metric.eval = c("TSS", "ROC"),
  metric.select.dataset= "validation",
  prob.mean = T,
  prob.cv = T,
  prob.ci = T,
  prob.ci.alpha = 0.05,
  prob.median = T,
  committee.averaging = T,
  prob.mean.weight = T,
  EMwmean.decay = 'proportional' )

bm_PlotEvalMean(bm.out = myBiomodEM)

bm_PlotResponseCurves(bm.out = myBiomodEM, 
                      models.chosen = get_built_models(myBiomodEM))
# Do model projection
myBiomodProj <- BIOMOD_Projection(
  bm.mod = myBiomodModelOut,  # Biomod2 model object
  new.env = myExpl1,  # New environmental data for projection
  proj.name = 'Salnepal',  # Name for the projection
  selected.models = 'all',  # Use all available models for projection
  binary.meth = "TSS",  # Binary method for thresholding (e.g., True Skill Statistic)
  build.clamping.mask = FALSE,  # Whether to build a clamping mask (typically set to FALSE)
  output.format = '.tif',# Output format for projected files (e.g., '.grd' for raster format)
  do.stack = FALSE
)
# Perform ensemble model projection using existing projection output
myBiomodEM_proj<- BIOMOD_EnsembleForecasting(
  bm.em = myBiomodEM,  # Biomod2 ensemble modeling output
  bm.proj = myBiomodProj,  # Projection output (already projected data)
  selected.models = 'all',  # Use all available models for ensemble projection
  proj.name = 'Salnepal',  # Name for the projection
  output.format = '.tif',
  metric.binary = "TSS", #without this function the binary map will not come
  binary.meth = "TSS",  # Binary method for thresholding (e.g., True Skill Statistic)
  do.stack = FALSE
)

# Get prediction for current habitat suitability
Proj <- get_predictions(myBiomodEM_proj) 
# Plot current habitat suitability map
plot(Proj)

# Create the individual raster objects
bio4 <- raster("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/future/ASCII/2070bio4_245.asc")
bio3 <- raster("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/future/ASCII/2070bio3_245.asc")
bio12 <- raster("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/future/ASCII/2070bio12_245.asc")
bio14 <- raster("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/future/ASCII/2070bio14_245.asc")
bio15 <- raster("F:/Khagendra (Ensemble analysis)/Worldclim data/Nepal/future/ASCII/2070bio15_245.asc")

# Stack the rasters together
myExpl2 <- stack(bio4, bio3, bio12, bio14, bio15, DEM, slope, aspect)

# Assign names to the layers in the stack
names(myExpl2) <- c("bio4", "bio3", "bio12", "bio14", "bio15", "dem", "slope", "aspect")

names(myExpl2)

# Do future model projection
myBiomodProj2 <- BIOMOD_Projection(
  bm.mod = myBiomodModelOut,  # Biomod2 model object
  new.env = myExpl2,  # New environmental data for projection
  proj.name = '2070',  # Name for the projection
  selected.models = 'all',  # Use all available models for projection
  binary.meth = "TSS",  # Binary method for thresholding (e.g., True Skill Statistic)
  build.clamping.mask = FALSE,  # Whether to build a clamping mask (typically set to FALSE)
  output.format = '.tif',# Output format for projected files (e.g., '.grd' for raster format)
  do.stack = FALSE
)

myBiomodEM_projfuture<- BIOMOD_EnsembleForecasting(
  bm.em = myBiomodEM,  # Biomod2 ensemble modeling output
  bm.proj = myBiomodProj2,  # Projection output (already projected data)
  selected.models = 'all',  # Use all available models for ensemble projection
  proj.name = '2070',  # Name for the projection
  output.format = '.tif',
  metric.binary = "TSS", #without this function the binary map will not come
  binary.meth = "TSS",  # Binary method for thresholding (e.g., True Skill Statistic)
  do.stack = FALSE
)

# Get prediction for future habitat suitability
Proj <- get_predictions(myBiomodEM_projfuture) 
plot(Proj)

[biomod2errorfiles.zip](https://github.com/user-attachments/files/15514101/biomod2errorfiles.zip)

**Comment by MayaGueguen** [comment]
Hello Joshi !

Thank you for posting with many details and informations :pray: 

I'm happy to say that thanks to your code, I found very quickly the problem :eyes: ... 
But really sorry about this error :disappointed: 

Actually, within [BIOMOD_Projection](https://biomodhub.github.io/biomod2/reference/BIOMOD_Projection.html) (and [BIOMOD_EnsembleForecasting](https://biomodhub.github.io/biomod2/reference/BIOMOD_EnsembleForecasting.html) too), we checked for similar variable names, but not for their order... And it seems that the order matters :face_in_clouds: 

I just pushed a commit on the github branch to correct for that, and we'll push it as soon as possible on CRAN.
Big apologies again :bowing_woman: 

Please let me know if it solves the problem.

Maya

**Comment by joshikp01** [comment]
Thank you Maya for the help.

Indeed the order was the problem !!

Grateful to you for solving my problem.

Sincerely,
Khagendra
